File: android/app/src/main/java/com/sausagemenu/app/OfflineMenuPlugin.java
package com.sausagemenu.app;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Rect;
import android.util.Base64;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.google.android.gms.tasks.Tasks;
import com.google.mlkit.common.model.DownloadConditions;
import com.google.mlkit.nl.languageid.LanguageIdentification;
import com.google.mlkit.nl.languageid.LanguageIdentificationOptions;
import com.google.mlkit.nl.languageid.LanguageIdentifier;
import com.google.mlkit.nl.translate.TranslateLanguage;
import com.google.mlkit.nl.translate.Translation;
import com.google.mlkit.nl.translate.Translator;
import com.google.mlkit.nl.translate.TranslatorOptions;
import com.google.mlkit.vision.common.InputImage;
import com.google.mlkit.vision.text.Text;
import com.google.mlkit.vision.text.TextRecognition;
import com.google.mlkit.vision.text.TextRecognizer;
import com.google.mlkit.vision.text.chinese.ChineseTextRecognizerOptions;
import com.google.mlkit.vision.text.devanagari.DevanagariTextRecognizerOptions;
import com.google.mlkit.vision.text.japanese.JapaneseTextRecognizerOptions;
import com.google.mlkit.vision.text.korean.KoreanTextRecognizerOptions;
import com.google.mlkit.vision.text.latin.TextRecognizerOptions;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Locale;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

@CapacitorPlugin(name = "OfflineMenu")
public class OfflineMenuPlugin extends Plugin {
    private static final Pattern TRAILING_PRICE = Pattern.compile(
        "((?:NT\\$|HK\\$|US\\$|A\\$|C\\$|[$€£¥₩฿₫₱₹₽])?\\s*\\d[\\d.,]*(?:\\s*(?:TWD|HKD|USD|JPY|KRW|EUR|GBP|THB|VND|PHP|RUB|INR|CNY|RMB|AUD|CAD))?)\\s*$",
        Pattern.CASE_INSENSITIVE
    );

    private final ExecutorService worker = Executors.newSingleThreadExecutor();

    @PluginMethod
    public void getStatus(PluginCall call) {
        JSObject result = new JSObject();
        result.put("available", true);
        result.put("engine", "ML Kit Text Recognition + On-device Translation");
        call.resolve(result);
    }

    @PluginMethod
    public void processMenu(PluginCall call) {
        JSArray images = call.getArray("images");
        String requestedTarget = call.getString("targetLanguage", "English");
        boolean allowsCellular = call.getBoolean("allowsCellularDownload", true);

        if (images == null || images.length() == 0) {
            call.reject("No menu images were provided.");
            return;
        }

        worker.execute(() -> {
            try {
                List<RecognizedLine> lines = new ArrayList<>();
                List<Object> imageValues = images.toList();
                for (int page = 0; page < imageValues.size(); page++) {
                    Bitmap bitmap = decodeImage(String.valueOf(imageValues.get(page)));
                    lines.addAll(recognizeBestScript(bitmap, page));
                }

                String fullText = joinOriginalText(lines);
                String sourceTag = identifyLanguage(fullText);
                String sourceLanguage = supportedLanguage(sourceTag);
                String targetLanguage = targetLanguage(requestedTarget);
                boolean modelDownloaded = false;

                if (!sourceLanguage.equals(targetLanguage)) {
                    TranslatorOptions options = new TranslatorOptions.Builder()
                        .setSourceLanguage(sourceLanguage)
                        .setTargetLanguage(targetLanguage)
                        .build();
                    Translator translator = Translation.getClient(options);
                    try {
                        DownloadConditions.Builder conditions = new DownloadConditions.Builder();
                        if (!allowsCellular) conditions.requireWifi();
                        try {
                            Tasks.await(translator.downloadModelIfNeeded(conditions.build()));
                        } catch (Exception downloadError) {
                            throw new IllegalStateException(
                                "This language pack is not downloaded yet. Please connect to the internet once and try again. / 尚未下載此語言包，請先連網一次後重試。",
                                downloadError
                            );
                        }
                        modelDownloaded = true;
                        for (RecognizedLine line : lines) {
                            if (!line.contentText.isEmpty()) {
                                line.translatedText = Tasks.await(translator.translate(line.contentText));
                            }
                        }
                    } finally {
                        translator.close();
                    }
                } else {
                    for (RecognizedLine line : lines) line.translatedText = line.contentText;
                }

                JSArray outputLines = new JSArray();
                for (RecognizedLine line : lines) outputLines.put(line.toJSObject());

                JSObject result = new JSObject();
                result.put("lines", outputLines);
                result.put("detectedLanguage", sourceTag);
                result.put("targetLanguage", targetLanguage);
                result.put("modelDownloaded", modelDownloaded);
                call.resolve(result);
            } catch (Exception error) {
                call.reject(error.getMessage() == null ? "Offline menu processing failed." : error.getMessage(), error);
            }
        });
    }

    private Bitmap decodeImage(String dataUrl) {
        int comma = dataUrl.indexOf(',');
        String encoded = comma >= 0 ? dataUrl.substring(comma + 1) : dataUrl;
        byte[] bytes = Base64.decode(encoded.getBytes(StandardCharsets.UTF_8), Base64.DEFAULT);
        Bitmap bitmap = BitmapFactory.decodeByteArray(bytes, 0, bytes.length);
        if (bitmap == null) throw new IllegalArgumentException("Unable to read the selected menu image.");
        return bitmap;
    }

    private List<RecognizedLine> recognizeBestScript(Bitmap bitmap, int page) throws Exception {
        InputImage image = InputImage.fromBitmap(bitmap, 0);
        List<TextRecognizer> recognizers = Arrays.asList(
            TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS),
            TextRecognition.getClient(new ChineseTextRecognizerOptions.Builder().build()),
            TextRecognition.getClient(new JapaneseTextRecognizerOptions.Builder().build()),
            TextRecognition.getClient(new KoreanTextRecognizerOptions.Builder().build()),
            TextRecognition.getClient(new DevanagariTextRecognizerOptions.Builder().build())
        );

        Text best = null;
        int bestScore = -1;
        try {
            for (TextRecognizer recognizer : recognizers) {
                Text candidate = Tasks.await(recognizer.process(image));
                int score = candidate.getText().replaceAll("\\s", "").length();
                if (score > bestScore) {
                    best = candidate;
                    bestScore = score;
                }
            }
        } finally {
            for (TextRecognizer recognizer : recognizers) recognizer.close();
        }

        List<RecognizedLine> output = new ArrayList<>();
        if (best == null) return output;
        for (Text.TextBlock block : best.getTextBlocks()) {
            for (Text.Line textLine : block.getLines()) {
                Rect box = textLine.getBoundingBox();
                if (box == null) continue;
                String original = textLine.getText().trim();
                if (original.isEmpty()) continue;
                Matcher priceMatcher = TRAILING_PRICE.matcher(original);
                String price = priceMatcher.find() ? priceMatcher.group(1).trim() : "";
                String content = priceMatcher.find(0) ? original.substring(0, priceMatcher.start()).trim() : original;
                if (content.isEmpty() && price.isEmpty()) continue;
                output.add(new RecognizedLine(
                    page,
                    original,
                    content,
                    price,
                    averageConfidence(textLine),
                    box.left / (double) bitmap.getWidth(),
                    box.top / (double) bitmap.getHeight(),
                    box.width() / (double) bitmap.getWidth(),
                    box.height() / (double) bitmap.getHeight()
                ));
            }
        }
        return output;
    }

    private double averageConfidence(Text.Line line) {
        double total = 0;
        int count = 0;
        for (Text.Element element : line.getElements()) {
            Float confidence = element.getConfidence();
            if (confidence != null) {
                total += confidence;
                count++;
            }
        }
        return count == 0 ? 0.8 : total / count;
    }

    private String joinOriginalText(List<RecognizedLine> lines) {
        StringBuilder text = new StringBuilder();
        for (RecognizedLine line : lines) text.append(line.originalText).append('\n');
        return text.toString();
    }

    private String identifyLanguage(String text) throws Exception {
        if (text.trim().isEmpty()) return "en";
        LanguageIdentifier identifier = LanguageIdentification.getClient(
            new LanguageIdentificationOptions.Builder().setConfidenceThreshold(0.25f).build()
        );
        try {
            String language = Tasks.await(identifier.identifyLanguage(text));
            return "und".equals(language) ? "en" : language;
        } finally {
            identifier.close();
        }
    }

    private String supportedLanguage(String tag) {
        String language = TranslateLanguage.fromLanguageTag(tag.toLowerCase(Locale.ROOT));
        return language == null ? TranslateLanguage.ENGLISH : language;
    }

    private String targetLanguage(String requested) {
        switch (requested) {
            case "繁體中文":
            case "繁體中文-HK": return TranslateLanguage.CHINESE;
            case "한국어": return TranslateLanguage.KOREAN;
            case "Français": return TranslateLanguage.FRENCH;
            case "Español": return TranslateLanguage.SPANISH;
            case "ไทย": return TranslateLanguage.THAI;
            case "Tagalog": return TranslateLanguage.TAGALOG;
            case "Tiếng Việt": return TranslateLanguage.VIETNAMESE;
            case "日本語": return TranslateLanguage.JAPANESE;
            case "Deutsch": return TranslateLanguage.GERMAN;
            case "Русский": return TranslateLanguage.RUSSIAN;
            case "Bahasa Indonesia": return TranslateLanguage.INDONESIAN;
            case "Polski": return TranslateLanguage.POLISH;
            case "Bahasa Melayu": return TranslateLanguage.MALAY;
            case "Italiano": return TranslateLanguage.ITALIAN;
            case "Português": return TranslateLanguage.PORTUGUESE;
            default: return TranslateLanguage.ENGLISH;
        }
    }

    @Override
    protected void handleOnDestroy() {
        worker.shutdownNow();
        super.handleOnDestroy();
    }

    private static final class RecognizedLine {
        final int page;
        final String originalText;
        final String contentText;
        final String protectedPrice;
        final double confidence;
        final double x;
        final double y;
        final double width;
        final double height;
        String translatedText = "";

        RecognizedLine(int page, String originalText, String contentText, String protectedPrice,
                       double confidence, double x, double y, double width, double height) {
            this.page = page;
            this.originalText = originalText;
            this.contentText = contentText;
            this.protectedPrice = protectedPrice;
            this.confidence = confidence;
            this.x = x;
            this.y = y;
            this.width = width;
            this.height = height;
        }

        JSObject toJSObject() {
            JSObject value = new JSObject();
            value.put("page", page);
            value.put("originalText", originalText);
            value.put("contentText", contentText);
            value.put("translatedText", translatedText);
            value.put("protectedPrice", protectedPrice);
            value.put("confidence", confidence);
            value.put("x", x);
            value.put("y", y);
            value.put("width", width);
            value.put("height", height);
            return value;
        }
    }
}

File: ios/App/App/OfflineMenuPlugin.swift
import Foundation
import UIKit
import Vision
import ImageIO
import NaturalLanguage
import Capacitor
import MLKitTranslate

private struct RecognizedMenuLine {
    let page: Int
    let originalText: String
    let contentText: String
    let protectedPrice: String?
    let confidence: Float
    let x: Double
    let y: Double
    let width: Double
    let height: Double
}

@objc(OfflineMenuPlugin)
public class OfflineMenuPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "OfflineMenuPlugin"
    public let jsName = "OfflineMenu"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "getStatus", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "processMenu", returnType: CAPPluginReturnPromise)
    ]

    private let priceRegex = try! NSRegularExpression(
        pattern: #"(?i)(?:\s|^)((?:NT\$|HK\$|US\$|A\$|C\$|[$€£¥₩฿₫₱₹₽])?\s*\d[\d.,]*(?:\s*(?:TWD|HKD|USD|JPY|KRW|EUR|GBP|THB|VND|PHP|RUB|INR|CNY|RMB|AUD|CAD|円|元|원))?)\s*$"#
    )

    @objc public func getStatus(_ call: CAPPluginCall) {
        call.resolve([
            "available": true,
            "engine": "Apple Vision OCR + ML Kit on-device translation",
            "minimumIOS": "15.5"
        ])
    }

    @objc public func processMenu(_ call: CAPPluginCall) {
        guard let images = call.getArray("images", String.self), !images.isEmpty else {
            call.reject("At least one menu image is required.")
            return
        }
        guard let targetTag = languageTag(for: call.getString("targetLanguage") ?? "繁體中文"),
              let targetLanguage = translateLanguage(for: targetTag) else {
            call.reject("The selected target language is not supported offline.")
            return
        }

        let allowsCellular = call.getBool("allowsCellularDownload") ?? false

        Task {
            do {
                var lines: [RecognizedMenuLine] = []
                for (page, image) in images.enumerated() {
                    lines.append(contentsOf: try recognizeText(in: image, page: page))
                }

                guard !lines.isEmpty else {
                    throw NSError(
                        domain: "OfflineMenu",
                        code: 1001,
                        userInfo: [NSLocalizedDescriptionKey: "No text was recognized in this image."]
                    )
                }

                let combinedText = lines.map(\.contentText).joined(separator: "\n")
                let sourceTag = detectLanguage(in: combinedText)
                guard let sourceLanguage = translateLanguage(for: sourceTag) else {
                    throw NSError(
                        domain: "OfflineMenu",
                        code: 1002,
                        userInfo: [NSLocalizedDescriptionKey: "The menu language is not supported by the offline model."]
                    )
                }

                let translatedLines: [[String: Any]]
                var downloadedModel = false
                if sourceLanguage == targetLanguage {
                    translatedLines = lines.map { dictionary(for: $0, translatedText: $0.contentText) }
                } else {
                    let options = TranslatorOptions(sourceLanguage: sourceLanguage, targetLanguage: targetLanguage)
                    let translator = Translator.translator(options: options)
                    do {
                        try await downloadModelIfNeeded(translator, allowsCellular: allowsCellular)
                    } catch {
                        throw NSError(
                            domain: "OfflineMenu",
                            code: 1004,
                            userInfo: [
                                NSLocalizedDescriptionKey: "This language pack is not downloaded yet. Please connect to the internet once and try again. / 尚未下載此語言包，請先連網一次後重試。"
                            ]
                        )
                    }
                    downloadedModel = true

                    var output: [[String: Any]] = []
                    output.reserveCapacity(lines.count)
                    for line in lines {
                        let translated: String
                        if line.contentText.isEmpty {
                            translated = ""
                        } else {
                            translated = try await translate(line.contentText, using: translator)
                        }
                        output.append(dictionary(for: line, translatedText: translated))
                    }
                    translatedLines = output
                }

                call.resolve([
                    "lines": translatedLines,
                    "detectedLanguage": sourceTag,
                    "targetLanguage": targetTag,
                    "modelDownloaded": downloadedModel
                ])
            } catch {
                call.reject(error.localizedDescription, nil, error)
            }
        }
    }

    private func recognizeText(in base64Image: String, page: Int) throws -> [RecognizedMenuLine] {
        let payload = base64Image.components(separatedBy: ",").last ?? base64Image
        guard let data = Data(base64Encoded: payload, options: .ignoreUnknownCharacters),
              let image = UIImage(data: data),
              let cgImage = image.cgImage else {
            throw NSError(
                domain: "OfflineMenu",
                code: 1003,
                userInfo: [NSLocalizedDescriptionKey: "The selected image could not be decoded."]
            )
        }

        let request = VNRecognizeTextRequest()
        request.recognitionLevel = .accurate
        request.usesLanguageCorrection = true
        request.minimumTextHeight = 0.008
        if #available(iOS 16.0, *) {
            request.automaticallyDetectsLanguage = true
        }

        let handler = VNImageRequestHandler(
            cgImage: cgImage,
            orientation: visionOrientation(for: image.imageOrientation),
            options: [:]
        )
        try handler.perform([request])
        let observations = request.results ?? []

        return observations.compactMap { observation in
            guard let candidate = observation.topCandidates(1).first else { return nil }
            let original = candidate.string.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !original.isEmpty else { return nil }
            let protected = splitProtectedPrice(from: original)
            let box = observation.boundingBox
            return RecognizedMenuLine(
                page: page,
                originalText: original,
                contentText: protected.content,
                protectedPrice: protected.price,
                confidence: candidate.confidence,
                x: box.minX,
                y: 1 - box.maxY,
                width: box.width,
                height: box.height
            )
        }
    }

    private func visionOrientation(for orientation: UIImage.Orientation) -> CGImagePropertyOrientation {
        switch orientation {
        case .up: return .up
        case .upMirrored: return .upMirrored
        case .down: return .down
        case .downMirrored: return .downMirrored
        case .left: return .left
        case .leftMirrored: return .leftMirrored
        case .right: return .right
        case .rightMirrored: return .rightMirrored
        @unknown default: return .up
        }
    }

    private func splitProtectedPrice(from text: String) -> (content: String, price: String?) {
        let range = NSRange(text.startIndex..<text.endIndex, in: text)
        guard let match = priceRegex.firstMatch(in: text, range: range),
              match.numberOfRanges > 1,
              let priceRange = Range(match.range(at: 1), in: text) else {
            return (text, nil)
        }

        let price = String(text[priceRange]).trimmingCharacters(in: .whitespacesAndNewlines)
        let contentEnd = Range(match.range(at: 0), in: text)?.lowerBound ?? text.endIndex
        let content = String(text[..<contentEnd]).trimmingCharacters(in: .whitespacesAndNewlines)
        return (content, price.isEmpty ? nil : price)
    }

    private func detectLanguage(in text: String) -> String {
        let recognizer = NLLanguageRecognizer()
        recognizer.processString(text)
        let detected = recognizer.dominantLanguage?.rawValue ?? "en"
        if translateLanguage(for: detected) != nil { return detected }
        let base = detected.split(separator: "-").first.map(String.init) ?? detected
        return translateLanguage(for: base) != nil ? base : "en"
    }

    private func translateLanguage(for languageTag: String) -> TranslateLanguage? {
        let normalized = languageTag
            .lowercased()
            .replacingOccurrences(of: "_", with: "-")
        let base = normalized.split(separator: "-").first.map(String.init) ?? normalized

        switch base {
        case "af": return .afrikaans
        case "sq": return .albanian
        case "ar": return .arabic
        case "be": return .belarusian
        case "bn": return .bengali
        case "bg": return .bulgarian
        case "ca": return .catalan
        case "zh": return .chinese
        case "hr": return .croatian
        case "cs": return .czech
        case "da": return .danish
        case "nl": return .dutch
        case "en": return .english
        case "eo": return .eperanto
        case "et": return .estonian
        case "fi": return .finnish
        case "fr": return .french
        case "gl": return .galician
        case "ka": return .georgian
        case "de": return .german
        case "el": return .greek
        case "gu": return .gujarati
        case "ht": return .haitianCreole
        case "he", "iw": return .hebrew
        case "hi": return .hindi
        case "hu": return .hungarian
        case "is": return .icelandic
        case "id": return .indonesian
        case "ga": return .irish
        case "it": return .italian
        case "ja": return .japanese
        case "kn": return .kannada
        case "ko": return .korean
        case "lv": return .latvian
        case "lt": return .lithuanian
        case "mk": return .macedonian
        case "ms": return .malay
        case "mt": return .maltese
        case "mr": return .marathi
        case "no": return .norwegian
        case "fa": return .persian
        case "pl": return .polish
        case "pt": return .portuguese
        case "ro": return .romanian
        case "ru": return .russian
        case "sk": return .slovak
        case "sl": return .slovenian
        case "es": return .spanish
        case "sw": return .swahili
        case "sv": return .swedish
        case "tl", "fil": return .tagalog
        case "ta": return .tamil
        case "te": return .telugu
        case "th": return .thai
        case "tr": return .turkish
        case "uk": return .ukrainian
        case "ur": return .urdu
        case "vi": return .vietnamese
        case "cy": return .welsh
        default: return nil
        }
    }

    private func languageTag(for displayName: String) -> String? {
        switch displayName {
        case "繁體中文", "繁體中文-HK": return "zh"
        case "English": return "en"
        case "한국어": return "ko"
        case "Français": return "fr"
        case "Español": return "es"
        case "ไทย": return "th"
        case "Tagalog": return "tl"
        case "Tiếng Việt": return "vi"
        case "日本語": return "ja"
        case "Deutsch": return "de"
        case "Русский": return "ru"
        case "Bahasa Indonesia": return "id"
        case "Polski": return "pl"
        case "Bahasa Melayu": return "ms"
        case "Italiano": return "it"
        case "Português": return "pt"
        default: return nil
        }
    }

    private func downloadModelIfNeeded(_ translator: Translator, allowsCellular: Bool) async throws {
        let conditions = ModelDownloadConditions(
            allowsCellularAccess: allowsCellular,
            allowsBackgroundDownloading: true
        )
        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<Void, Error>) in
            translator.downloadModelIfNeeded(with: conditions) { error in
                if let error = error {
                    continuation.resume(throwing: error)
                } else {
                    continuation.resume(returning: ())
                }
            }
        }
    }

    private func translate(_ text: String, using translator: Translator) async throws -> String {
        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<String, Error>) in
            translator.translate(text) { translatedText, error in
                if let error = error {
                    continuation.resume(throwing: error)
                } else if let translatedText = translatedText {
                    continuation.resume(returning: translatedText)
                } else {
                    continuation.resume(returning: text)
                }
            }
        }
    }

    private func dictionary(for line: RecognizedMenuLine, translatedText: String) -> [String: Any] {
        var value: [String: Any] = [
            "page": line.page,
            "originalText": line.originalText,
            "contentText": line.contentText,
            "translatedText": translatedText,
            "confidence": Double(line.confidence),
            "x": line.x,
            "y": line.y,
            "width": line.width,
            "height": line.height
        ]
        if let price = line.protectedPrice {
            value["protectedPrice"] = price
        }
        return value
    }
}
